### PR TITLE
feat(WebexMeetingSharing): add screen share meeting control and stories

### PR DESCRIPTION
**Pull request by yusenjeng**
To add screen share meeting control, remote sharing video in the WebexRemoteMedia, related json adapter and stories.

### CODE REVIEW

File: src/components/hooks/useShareControl.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import {AdapterContext} from '../../components/';
+
+/**
+ * Custom hook that returns the display state of the meeting share control.
+ *
+ * @param {string} meetingID  ID of the meeting for which to get data
+ * @returns {shareControlDisplay}  Data state of the share control
+ */
+export default function useShareControl(meetingID) {
```
Comment (lalli-flores) [inline src/components/hooks/useShareControl.js:11]: Not sure I understand the purpose of this hook? `screen-share` should work like any other control and be useable through the [existing `useMeetingControl` hook](https://github.com/webex/components/blob/master/src/components/hooks/useMeetingControl.js)

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js [hunk 2]
```diff
@@ -17,15 +16,41 @@ import './WebexRemoteMedia.scss';
...
   };
 
+  const isLocalSharingInactive = shareControl.state === 'inactive';
+
+  //
+  // Switch the visibility of the remoteShare video by checking its readyState
+  // Note: Sharing detection can be removed from component, once the sdk provides
+  //       the Start/Stop events about the remote sharing.
+  //
+  const detectRemoteSharing = () => {
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.js:37]: This is not necessary in this component. `WebexMeeting` component handles the layering of local media + remote media + meeting controls. `WebexRemoteMedia` should only be concerned with displaying remote streams.
Do you have suggestions on how to make that clearer in the docs/repo?

File: src/data/meetings.json [hunk 3]
```diff
@@ -120,13 +120,31 @@
   "localMedia": {
     "ID": "localMedia",
     "title": "Ad Hoc Meeting Space",
-    "localVideo": {},
+    "localVideo": {
+      "muted": false
```
Comment (lalli-flores) [inline src/data/meetings.json:124]: Is there a reason for adding the `muted` property?

File: src/adapters/MeetingsJSONAdapter.js [hunk 4]
```diff
@@ -157,6 +164,17 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
       /* eslint-enable no-confusing-arrow */
     );
 
+    const shareEvents$ = fromEvent(document, SHARE_CONTROL).pipe(
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js:167]: I see that you're adding the `remoteShare` whenever the share control is activated, however, I think this should follow the pattern of the other streams, where you add local/remote screen share stream based on the JSON data.

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js (hunk 2 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.js:37]: This is the tricky part. 

The idea of `if (remoteShare) then display <video>` doesn't work for the sdk, because the remoteShare/remoteVideo stream object will be always true value during the entire meeting. The difference is that it receives data or not, and the stream objects were always there.

And, unfortunately, the sdk doesn't provide the remote start/stop events as well, so the adapter can't provide the states of those remote media. That means UI components have no way to know the remote start/stop sharing and cannot do proper UI changes.

This is a sdk design problem plus a missing sdk feature.

But we still found a way to detect the remote event from the DOM video element as a workaround, that's detectRemoteSharing(). DOM element was managed by React, so the code exists in the component.

We can surely remove the detection function, but the UI will lose ability to provide feedbacks to the users about the changes of remote steam (with the current sdk).

File: src/adapters/MeetingsJSONAdapter.js [hunk 5]
```diff
@@ -207,6 +225,24 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
     return stream;
   }
 
+  /**
+   * Returns a MediaStream object obtained from user local sharing.
+   * @param {MediaStreamConstraints} displayMediaOptions  an object specifying the types of the media to request
+   * @returns {MediaStream}
+   */
+  async getSharingStream(displayMediaOptions) {
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/adapters/MeetingsJSONAdapter.js:234]: got it

File: src/adapters/MeetingsJSONAdapter.js [hunk 6]
```diff
@@ -438,6 +474,77 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
+      if (meeting) {
+        observer.next(meeting.localShare ? active : inactive);
+      } else {
+        observer.error(new Error(`Could not find meeting with ID "${ID}"`));
+      }
+
+      observer.complete();
+    });
+
+    const muteEvent$ = fromEvent(document, SHARE_CONTROL).pipe(
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/adapters/MeetingsJSONAdapter.js:541]: agree

File: src/components/WebexMeeting/WebexMeeting.stories.js [hunk 7]
```diff
@@ -16,11 +16,33 @@ stories.add('default', () => (
 ));
 
 stories.add('custom controls', () => {
-  const controls = (isActive) => (isActive ? ['leave-meeting'] : ['join-meeting']);
+  const controls = (isActive) => {
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/components/WebexMeeting/WebexMeeting.stories.js:19]: No strong reason for that. Will rollback and focus on the sharing part.

File: src/data/meetings.json (hunk 3 above)
Comment (yusenjeng) [reply src/data/meetings.json:124]: No strong reason for that. Will rollback and focus on the sharing part.

File: src/components/hooks/useMeeting.js [hunk 8]
```diff
@@ -16,6 +16,7 @@ export default function useMeeting(meetingID) {
     localVideo: null,
     remoteAudio: null,
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/components/hooks/useMeeting.js:17]: It's about UX.  There are still many sharing bugs inside the sdk. It fails often, so we are not sure to display or not to display the localShare on the sharing sender's screen at this point.

I prefer leave the application of localShare to the next version.

[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/components/hooks/useMeeting.js:17]: Sounds good 👍🏻

File: src/data/meetings.json [hunk 9]
```diff
@@ -122,11 +122,27 @@
...
+    },
     "remoteAudio": null,
     "localShare": null,
     "remoteShare": null
   },
+  "localMediaWithSharing": {
+    "ID": "localMediaWithSharing",
+    "title": "Ad Hoc Meeting Space",
+    "localVideo": {
```
Comment (lalli-flores) [inline src/data/meetings.json:135]: Since this is supposed to represent a meeting with screen share, the attributes that represent a display media stream should not be `null`, and instead be an object to indicate that "they exist". Like so:
```suggestion
    "localVideo": {},
    "remoteVideo": {},
    "localAudio": {},
    "remoteAudio": {},
    "localShare": {},
    "remoteShare": null
```

In this case, having an object in the `localShare` attribute indicates that the user is sharing their screen. Conversely, if `remoteShare` was an object `{}` instead of `null`, while `localShare` is `null`, this would represent a meeting where a remote participant is sharing their screen.

File: src/components/hooks/useShareControl.js (hunk 1 above)
Comment (yusenjeng) [reply src/components/hooks/useShareControl.js:11]: Will reuse the useMeetingControl().

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js (hunk 2 above)
Comment (lalli-flores) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.js:37]: A couple things here:
* The JSON adapter code seems to be making changes to `localShare` only, so why is the remote media component being changed?
* I think there might be some rework that needs to happen on the SDK adapter if what you're saying is how things are currently done. SDK issues/design should not be bubbling up to the components, thats why the adapter is there. The SDK adapter should do what it needs to to make sure that if there is something being shared, then it returns the right stream, whereas if nothing is being shared, then the attribute should be `null`. This is why in the SDK adapter there are two objects `meeting` which is the component systems state, and `sdkMeeting`, which is the state of how the SDK handles meetings; the job of the adapter to to translate `sdkMeeting` into `meeting` so that the components don't have to know any of those implementation details.

File: src/data/meetings.json [hunk 10]
```diff
@@ -122,11 +122,27 @@
     "title": "Ad Hoc Meeting Space",
     "localVideo": {},
     "remoteVideo": null,
-    "localAudio": {},
+    "localAudio": {
+      "muted": false
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/data/meetings.json:126]: sure

File: src/data/meetings.json (hunk 9 above)
Comment (yusenjeng) [reply src/data/meetings.json:135]: How about to set both the `localShare` and the `remoteShare` to `null`?  This scenario requires the user to click on the sharing button by themself. The sharing video isn't acquired automatically on starting.

File: src/adapters/MeetingsJSONAdapter.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/adapters/MeetingsJSONAdapter.js]: agree

File: src/components/hooks/useStream.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ export default function useStream(stream) {
   useEffect(() => {
     const mediaElement = streamRef.current;
 
-    if (stream instanceof MediaStream) {
+    if (!!mediaElement && stream instanceof MediaStream) {
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/components/hooks/useStream.js:15]: ok

File: src/data/meetings.json (hunk 9 above)
Comment (lalli-flores) [reply src/data/meetings.json]: I'm ok with that, if it's not necessary to add the share stream. In that case, I would look to the existing objects to see if you can reuse any of them instead of adding new one

Comment (yusenjeng) [reply src/data/meetings.json]: good idea. Will reuse "remote&localMedia" instead of the new one.

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js [hunk 12]
```diff
@@ -17,15 +16,40 @@ import './WebexRemoteMedia.scss';
...
+    if (!shareRef.current) {
+      return false;
+    }
+
+    return shareRef.current.readyState > 1;
+  };
+
+  const remoteShareClasses = {
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.js:44]: Are these necessary? If they are, this should probably be unified at the component level in [`mainClasses`](https://github.com/webex/components/blob/master/src/components/WebexRemoteMedia/WebexRemoteMedia.js#L24)

File: src/components/WebexMeeting/WebexMeeting.stories.js [hunk 13]
```diff
@@ -16,11 +16,33 @@ stories.add('default', () => (
...
+  };
 
   return (
     <WebexDataProvider adapter={webexAdapter}>
       <WebexMeeting meetingDestination="localMedia" controls={controls} />
     </WebexDataProvider>
   );
 });
+
+stories.add('custom controls with sharing', () => {
```
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/components/WebexMeeting/WebexMeeting.stories.js:28]: I'd still like to see it done this way

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js [hunk 14]
```diff
@@ -40,7 +64,13 @@ export default function WebexRemoteMedia({className, meetingID}) {
               <div>Connecting</div>
             </Badge>
           ) : null}
-          {remoteVideo ? <video ref={videoRef} playsInline autoPlay /> : null}
+
+          {remoteVideo ? <video ref={videoRef} playsInline autoPlay name="remote-cam" /> : null}
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.js:68]: I don't think [`name` is an attribute of `video`](https://developer.mozilla.org/en-US/docs/Web/API/Element/name) elements

```suggestion
          {remoteVideo ? <video ref={videoRef} playsInline autoPlay /> : null}
```

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js [hunk 15]
```diff
@@ -40,7 +64,13 @@ export default function WebexRemoteMedia({className, meetingID}) {
...
+
+          {remoteShare && isLocalSharingInactive ? (
+            <video ref={shareRef} playsInline autoPlay name="remote-share" className={classNames(remoteShareClasses)} />
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.js:71]: I don't think [`name` is an attribute of `video`](https://developer.mozilla.org/en-US/docs/Web/API/Element/name) elements
```suggestion
            <video ref={shareRef} playsInline autoPlay className={classNames(remoteShareClasses)} />
```

File: src/components/WebexRemoteMedia/WebexRemoteMedia.scss [hunk 16]
```diff
@@ -26,5 +26,17 @@
   video {
     height: 100%;
     width: 100%;
+
+    &.share {
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.scss:30]: Is something using this?

File: package.json [hunk 17]
```diff
@@ -39,6 +39,7 @@
     "@juggle/resize-observer": "^3.0.2",
     "@momentum-ui/react": "^23.2.1",
     "@webex/component-adapter-interfaces": "^1.9.0",
+    "@webex/sdk-component-adapter": "file:../sdk-adapter-official",
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply package.json:42]: It shouldn't be there.

File: src/components/WebexInMeeting/WebexInMeeting.js [hunk 18]
```diff
@@ -27,7 +27,9 @@ export default function WebexInMeeting({meetingID}) {
 
   return (
     <div ref={meetingRef} className={classNames(mainClasses)}>
-      <WebexRemoteMedia className="remote-media-in-meeting" meetingID={meetingID} />
+      <MeetingContext.Provider value={meetingID}>
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/components/WebexInMeeting/WebexInMeeting.js:30]: The useMeetingControl hook needs this to provide the context. <WebexMeetingControls> is a similar scenario. Its sub-components also use useMeetingControl

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js (hunk 14 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.js:68]: This custom attribute was used to differentiate the two video element when they appeared at the same time. Removing them is fine.

File: src/components/WebexRemoteMedia/WebexRemoteMedia.scss (hunk 16 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.scss:30]: Yes, it's part of remoteShareClasses used to control the visibility of sharing video (not the cam video).

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js (hunk 12 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.js:44]: It is used to stylized one single video element inside the `<WebexRemoteMedia>`, not the whole component. However, the video element itself is not big enough to be a standalone component.

File: src/components/WebexMeetingControl/WebexMeetingControl.stories.js [hunk 19]
```diff
@@ -49,3 +49,13 @@ stories.add('disabled icon', () => (
     </WebexMeetingControls>
   </WebexDataProvider>
 ));
+
+stories.add('multiple icons', () => (
+  <WebexDataProvider adapter={webexAdapter}>
+    <WebexMeetingControls meetingID="localMediaWithSharing">
```
Comment (lalli-flores) [inline src/components/WebexMeetingControl/WebexMeetingControl.stories.js:55]: This meeting ID no longer exists, so the story is not displaying properly

File: src/components/WebexInMeeting/WebexInMeeting.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (lalli-flores) [reply src/components/WebexInMeeting/WebexInMeeting.js:30]: Maybe I haven't had enough coffee yet but the `useMeetingControl` hook fetches the appropriate context already, why is this necessary?

[the remark this answers is not in the document]
Comment (yusenjeng) [reply src/components/WebexInMeeting/WebexInMeeting.js]: The MeetingContext will be removed for simplification.

File: src/components/WebexRemoteMedia/WebexRemoteMedia.js (hunk 12 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.js]: The `remoteShareClasses` will be removed. Instead, we choose a better way to do split screen.

File: src/components/WebexRemoteMedia/WebexRemoteMedia.scss (hunk 16 above)
Comment (yusenjeng) [reply src/components/WebexRemoteMedia/WebexRemoteMedia.scss]: This part is also changed for the removal of `remoteShareClasses`.

File: src/components/index.js [hunk 20]
```diff
@@ -2,11 +2,11 @@ export {default as WebexAvatar} from './WebexAvatar/WebexAvatar';
...
 export {default as WebexDataProvider, AdapterContext} from './WebexDataProvider/WebexDataProvider';
-export {default as WebexInMeeting} from './WebexInMeeting/WebexInMeeting';
-export {default as WebexInterstitialMeeting} from './WebexInterstitialMeeting/WebexInterstitialMeeting';
 export {default as WebexLocalMedia} from './WebexLocalMedia/WebexLocalMedia';
 export {default as WebexMeeting} from './WebexMeeting/WebexMeeting';
 export {default as WebexMeetingControl} from './WebexMeetingControl/WebexMeetingControl';
 export {default as WebexMeetingControls, MeetingContext} from './WebexMeetingControl/WebexMeetingControls';
 export {default as WebexMeetingInfo} from './WebexMeetingInfo/WebexMeetingInfo';
 export {default as WebexRemoteMedia} from './WebexRemoteMedia/WebexRemoteMedia';
+export {default as WebexInMeeting} from './WebexInMeeting/WebexInMeeting';
```
Comment (lalli-flores) [inline src/components/index.js:11]: Can you revert this change? We wanted to keep the exports alphabetica

File: package.json [hunk 21]
```diff
@@ -149,4 +149,4 @@
       "commit-msg": "commitlint -E HUSKY_GIT_PARAMS"
     }
   }
-}
+}
```
Comment (lalli-flores) [inline package.json:152]: Annoying change but can you add a new line at the end?

File: package-lock.json [hunk 22]
```diff
@@ -22227,4 +22227,4 @@
       }
     }
   }
-}
+}
```
Comment (lalli-flores) [inline package-lock.json:22230]: Annoying change but can you add a new line at the end?

File: src/adapters/MeetingsJSONAdapter.js [hunk 23]
```diff
@@ -438,6 +474,78 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
 
+  /**
+   * Handles the starting and stopping of the local screen capture media stream.
+   *
+   * @param {string} ID ID of the meeting for which to update display
+   * @memberof MeetingJSONAdapter
+   * @private
+   */
+  async handleLocalShare(ID) {
+    console.log('handleLocalShare', ID);
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js:485]: Can you remove this please?

File: src/components/index.js (hunk 20 above)
Comment (yusenjeng) [reply src/components/index.js:11]: ok

File: package.json (hunk 21 above)
Comment (yusenjeng) [reply package.json]: sure

File: package-lock.json (hunk 22 above)
Comment (yusenjeng) [reply package-lock.json]: sure

File: src/adapters/MeetingsJSONAdapter.js (hunk 23 above)
Comment (yusenjeng) [reply src/adapters/MeetingsJSONAdapter.js]: already removed

File: src/components/WebexMeetingControl/WebexMeetingControl.stories.js (hunk 19 above)
Comment (yusenjeng) [reply src/components/WebexMeetingControl/WebexMeetingControl.stories.js]: use localMedia instead

File: package-lock.json [hunk 24]
```diff
@@ -3682,6 +3682,18021 @@
       "integrity": "sha512-LfmH444YDTmvc0zqqOm55a1WxVk89XMh/Eq/Sv0YaW8CJCp2m1Ib3jKLl9+Qu9F3dSHpiCxHfjW+k++RX3NwVQ==",
       "dev": true
     },
+    "@webex/sdk-component-adapter": {
```
[the remark this answers is not in the document]
Comment (yusenjeng) [reply package-lock.json]: it's done

File: src/components/WebexRemoteMedia/WebexRemoteMedia.test.js [hunk 25]
```diff
@@ -4,6 +4,7 @@ import WebexRemoteMedia from './WebexRemoteMedia';
 
 jest.mock('../hooks/useMeeting');
 jest.mock('../hooks/useStream');
+jest.mock('../hooks/useMeetingControl');
```
Comment (lalli-flores) [inline src/components/WebexRemoteMedia/WebexRemoteMedia.test.js:7]: Can you remove this? This hook is no longer used in `WebexRemoteMedia`
